File: java/lesson 5/Trie.java
// https://leetcode.com/problems/implement-trie-prefix-tree/

public class Trie {
    private TrieNode root = new TrieNode();
    
    public void insert(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (node.get(c) == null) {
                node.set(c, new TrieNode());
            }
            node = node.get(c);
        }
        node.isWord = true;
    }
    
    public boolean search(String word) {
        TrieNode node = findNode(word);
        return node != null && node.isWord;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode node = findNode(prefix);
        return node != null;
    }
    
    private TrieNode findNode(String prefix) {
        TrieNode node = root;
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            node = node.get(c);
            if (node == null) return null;
        }
        return node;
    }
    
    private class TrieNode {
        private TrieNode[] children = new TrieNode[26]; // a to z
        public boolean isWord;
        
        public TrieNode get(Character c) {
            return children[c - 'a'];
        }
        
        public void set(Character c, TrieNode value) {
            children[c - 'a'] = value;
        }
    }
}

File: csharp/lesson 5/Trie.cs
// https://leetcode.com/problems/implement-trie-prefix-tree/

# nullable enable
public class Trie
{
    private TrieNode root = new TrieNode();
    
    public void Insert(String word) 
    {
        var node = root;
        foreach (var c in word) 
        {
            if (node[c] == null)
            {
                node[c] = new TrieNode();
            }
            node = node[c];
        }
        node.IsWord = true;
    }
    
    public bool Search(String word)
    {
        var node = root;
        foreach (var c in word)
        {
            if (node[c] == null) return false;
            node = node[c];
        }
        return node.IsWord;
    }
    
    public bool StartsWith(String prefix) {
        var node = root;
        foreach (var c in prefix) 
        {
            if (node[c] == null) return false;
            node = node[c];
        }
        return true;
    }
    
    private class TrieNode {
        private TrieNode[] children = new TrieNode[26]; // a to z
        public bool IsWord { get; set; }
        
        public TrieNode? this[char c]
        {
            get => children[c - 'a'];
            set => children[c - 'a'] = value;
        }
    }
}
